File: src/main/java/us/hervalicio/ai/neural/NumericDistribution.java
package us.hervalicio.ai.neural;

import java.util.Random;

/**
 * Created by herval on 10/30/15.
 */
public class NumericDistribution {
    private Random random = new Random(12345);

    /**
     * Given a probability distribution over discrete classes, sample from the distribution
     * and return the generated class index.
     *
     * @param distribution Probability distribution over classes. Must sum to 1.0
     */
    public int sample(double[] distribution) {
        double d = random.nextDouble();
        double sum = 0.0;
        for (int i = 0; i < distribution.length; i++) {
            sum += distribution[i];
            if (d <= sum) return i;
        }
        //Should never happen if distribution is a valid probability distribution
        throw new IllegalArgumentException("Distribution is invalid? d=" + d + ", sum=" + sum);
    }
}

File: src/main/scala/us/hervalicio/ai/lstm/NumericDistribution.scala
package us.hervalicio.ai.lstm

import java.util.Random

class NumericDistribution {
  private val random = new Random()

  /**
    * Given a probability distribution over discrete classes, sample from the distribution
    * and return the generated class index.
    *
    * @param distribution Probability distribution over classes. Must sum to 1.0
    */
  def sample(distribution: Array[Double]): Int = {
    val d: Double = random.nextDouble
    var sum = 0.0
    distribution.zipWithIndex.foreach { case(n, i) =>
        sum += n
        if(d <= sum) {
          return i
        }
    }
    throw new IllegalArgumentException("Distribution is invalid? d=" + d + ", sum=" + sum)
  }
}